### PR TITLE
Java: CWE-297 insecure JavaMail SSL configuration

### PR DESCRIPTION
**Pull request by luchua-bc**
JavaMail is widely used in Java applications for sending emails. There are some other popular third-party libraries like Apache Commons Email which are built on JavaMail and facilitate the integration. Authenticated mail sessions require user credentials and mail sessions can require SSL/TLS authentication. It is a common security vulnerability that host-specific certificate data is not validated or is incorrectly validated. Failing to validate the certificate makes the SSL session susceptible to a man-in-the-middle attack.

This query validates configuration of both JavaMail and SimpleMail. I've tested the query against some GitHub projects, and two of them are:
1. [Apache Log4J project using JavaMail](https://github.com/apache/logging-log4j2). The version 2.13.1 has [Add support for specifying an SSL configuration for SmtpAppender (CVE-2020-9488)](https://issues.apache.org/jira/browse/LOG4J2-2819) (not reported by me)
2. [cronomst/simplemail project using Apache Commons Email](https://github.com/cronomst/simplemail)

Please consider to merge the PR. Thanks.

@luchua-bc

### CODE REVIEW

File: java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+ * 1. Set the mail.smtp.auth property to provide the SMTP Transport with a username and password when connecting to the SMTP server or
+ *    set the mail.smtp.ssl.socketFactory/mail.smtp.ssl.socketFactory.class property to create an SMTP SSL socket.
+ * 2. No mail.smtp.ssl.checkserveridentity property is enabled.
+ */
+predicate isInsecureMailPropertyConfig(VarAccess propertiesVarAccess) {
+  exists(MethodAccess ma |
+    ma.getMethod() instanceof SetPropertyMethod and
+    ma.getQualifier() = propertiesVarAccess.getVariable().getAnAccess() and
+    (
+      getStringValue(ma.getArgument(0)).indexOf(".auth") != -1 and //mail.smtp.auth
```
Comment (aschackmull) [inline java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql:36]: `indexOf` can never give `-1`, if the string doesn't contain the substring `".auth"` then `indexOf` will simply yield zero results, so `x.indexOf("foo") != -1` is equivalent to `exists(x.indexOf("foo"))`, which in turn can be expressed simpler with `.matches("%foo%")`.
```suggestion
      getStringValue(ma.getArgument(0)).matches("%.auth%") and //mail.smtp.auth
```

File: java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+      getStringValue(ma.getArgument(1)) = "true"
+      or
+      getStringValue(ma.getArgument(0)).indexOf(".socketFactory") != -1 //mail.smtp.socketFactory or mail.smtp.socketFactory.class
```
Comment (aschackmull) [inline java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql:39]: same

File: java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    )
+  ) and
+  not exists(MethodAccess ma |
+    ma.getMethod() instanceof SetPropertyMethod and
+    ma.getQualifier() = propertiesVarAccess.getVariable().getAnAccess() and
+    (
+      getStringValue(ma.getArgument(0)).indexOf(".ssl.checkserveridentity") != -1 and //mail.smtp.ssl.checkserveridentity
```
Comment (aschackmull) [inline java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql:46]: same

File: java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+/**
+ * Helper method to get string value of an argument
+ */
+string getStringValue(Expr expr) {
+  result = expr.(StringLiteral).getRepresentedString()
+  or
+  exists(Variable v | expr = v.getAnAccess() |
+    result = getStringValue(v.getInitializer().(CompileTimeConstantExpr))
```
Comment (aschackmull) [inline java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql:59]: This gets the value from the initializer even if the variable is subsequently modified.  I assume you meant to restrict this to the value of `CompileTimeConstantExpr`s?  In that case, both this case and the base case above can be replaced with `result = expr.(CompileTimeConstantExpr).getStringValue()`.

File: java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+  result = getStringValue(expr.(AddExpr).getLeftOperand())
+  or
+  result = getStringValue(expr.(AddExpr).getRightOperand())
+}
+
+/**
+ * The JavaMail session class `javax.mail.Session`
+ */
+class MailSession extends RefType {
+  MailSession() { this.getQualifiedName() = "javax.mail.Session" }
```
Comment (aschackmull) [inline java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql:71]: For matching a specific type we prefer `hasQualifiedName` over `getQualifiedName` as the latter performs a lot more string operations under the hood.
```suggestion
  MailSession() { this.hasQualifiedName("javax.mail", "Session") }
```

File: java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+/**
+ * The class of Apache SimpleMail
+ */
+class SimpleMail extends RefType {
+  SimpleMail() { this.getQualifiedName() = "org.apache.commons.mail.SimpleEmail" }
```
Comment (aschackmull) [inline java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql:78]: same comment

File: java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    ma.getMethod().hasName("setSSLCheckServerIdentity") and
+    ma.getArgument(0).(BooleanLiteral).getBooleanValue() = true
+  )
+}
+
+from MethodAccess ma
+where
+  ma.getMethod().getDeclaringType() instanceof MailSession and
+  ma.getMethod().getName() = "getInstance" and
+  isInsecureMailPropertyConfig(ma.getArgument(0).(VarAccess))
```
Comment (aschackmull) [inline java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql:104]: No need to cast to `VarAccess` here as that's already done implicitly by being an argument to `isInsecureMailPropertyConfig`.

File: java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+  or
+  enableTLSWithSimpleMail(ma) and hasNoCertCheckWithSimpleMail(ma.getQualifier().(VarAccess))
```
Comment (aschackmull) [inline java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql:106]: Same here. The `.(VarAccess)` cast is superfluous.

File: java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql (hunk 1 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql]: Thanks @aschackmull . I've made the change in the new commit.

File: java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql (hunk 2 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql]: Same as above.

File: java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql (hunk 3 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql]: Same as above.

File: java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql (hunk 4 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql]: Thanks @aschackmull . I've incorporated the change to simplify the query.

File: java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql (hunk 5 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql]: Thanks @aschackmull for the suggestion. Glad to know the difference and I will use hasQualifiedName down the road.

File: java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql (hunk 6 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql]: Same as above. Thanks.

File: java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql (hunk 7 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql]: Thanks. Change was made.

File: java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql (hunk 8 above)
Comment (luchua-bc) [reply java/ql/src/experimental/Security/CWE/CWE-297/InsecureJavaMail.ql]: Same as above.

**Comment by aschackmull** [comment]
Looks like the autoformat PR check is failing.  Could you autoformat the QL code, please?

**Comment by luchua-bc** [comment]
Sorry @aschackmull . I did autoformat the QL code in the previous commit then removed the ending blank line since it caused an autoformat issue with a previous PR. 

I just committed a new version with the autoformatted code from the CodeQL plugin of VS Code on Mac, which contains an ending blank line. Let's see whether this helps. If it doesn't, I will try to remove the ending line in GitHub GUI directly.

**Comment by aschackmull** [comment]
The last character in a text file should be a newline character (and the github ui will show a small icon when that isn't the case). The trouble is that some editors don't follow this convention and show this as a blank line at the end of the file (Eclipse has this behaviour), so this sometimes results in some confusion about whether or not there is a blank line at the end of a file.

In any case, the autoformat check passed and now continued to run the newly added test (we greatly appreciate tests for new queries, so thanks for adding one!). However, the test failed, since such tests need to be self-contained and cannot rely on external libraries. We usually handle this by stubbing any needed dependencies (see the `java/ql/test/stubs` dir).

**Comment by luchua-bc** [comment]
Thanks @aschackmull . I've added stub classes of Apache Commons Email for the test case. Hope the Java check can pass now.

**Comment by aschackmull** [comment]
The added `Email.java` doesn't look like a stub, but rather a complete copy. We prefer stubs. Also, the test doesn't quite work yet as there are some more dependencies missing.  You can try to execute the test locally to verify that it works through the Test Explorer in VSCode.

**Comment by luchua-bc** [comment]
@aschackmull . Thanks for the advice, I've added all dependencies as class stubs. I used to use binary JAR files directly during my local testing. Now the Java check shall pass.

**Pull request merged**